### PR TITLE
[7/n] Migrate pos_encoding and norm kernels to libtorch stable ABI (continued)

### PR DESCRIPTION
**Pull request by cleonard530**
This is a continuation of the PR #38783

cc @janeyx99

-------------------------------------------------------------------------------------------------------------------

## Purpose
~~Stacked on https://github.com/vllm-project/vllm/pull/38757, commits to review https://github.com/vllm-project/vllm/pull/38783/changes/deea6618c38afb4735b442c61e2697c273654292..8754a4250584115db08113e0889313c939d85eb6~~

Note: some declarations are not deleted from csrc/ops.h despite being moved to csrc/libtorch_stable/ops.h. This is because the CPU build also uses these declarations. These are

 - Layernorm kernels: rms_norm, fused_add_rms_norm
  - Pos encoding kernels: rotary_embedding 
  
## Test Plan

```
pytest tests/kernels/core/test_pos_encoding.py
pytest tests/kernels/core/test_fused_qk_norm_rope.py
pytest tests/kernels/core/test_layernorm.py
pytest tests/kernels/core/test_fused_quant_layernorm.py
```

## Test Result

Screenshot 2026-04-02 at 2 45 33 PM
Screenshot 2026-04-02 at 2 45 01 PM

Screenshot 2026-04-02 at 2 46 02 PM

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft in the [Google Doc](https://docs.google.com/document/d/1YyVqrgX4gHTtrstbq8oWUImOyPCKSGnJ7xtTpmXzlRs/edit?tab=t.0).
</details>

------------------------------------------------------------------------------------------------------
**Differences between this and PR #38783**
- The merge conflicts that were fixed. These are pointed out in the comments below. 
- Added comments in csrc/ops.h mentioning that some of the ops where in there as well as the libtorch_stable/ops.h because the CPU build still needed them.

**Migration progress using the Audit Python extension [torch-abi-audit](https://pypi.org/project/torch-abi-audit/):**

  **Before update:**
```
      -- extensions --
      [UNSTABLE] [abi3-ok               ] _C.abi3.so  (stable_shim=0, unstable=101)
      [STABLE  ] [abi3-ok               ] _C_stable_libtorch.abi3.so  (stable_shim=78, unstable=0)
      [UNSTABLE] [abi3-ok               ] _moe_C.abi3.so  (stable_shim=0, unstable=82)
      [NO-TORCH] [abi3-ok               ] cumem_allocator.abi3.so
      [NO-TORCH] [abi3-ok               ] spinloop.abi3.so
      [UNSTABLE] [uses-private-api      ] third_party/deep_gemm/_C.cpython-312-x86_64-linux-gnu.so  (stable_shim=0, unstable=59)
      [UNSTABLE] [abi3-ok               ] vllm_flash_attn/_vllm_fa2_C.abi3.so  (stable_shim=0, unstable=85)
      [UNSTABLE] [abi3-ok               ] vllm_flash_attn/_vllm_fa3_C.abi3.so  (stable_shim=0, unstable=81)
```

  **After update:**
```
      -- extensions --
      [UNSTABLE] [abi3-ok               ] _C.abi3.so  (stable_shim=0, unstable=99)
      [STABLE  ] [abi3-ok               ] _C_stable_libtorch.abi3.so  (stable_shim=78, unstable=0)
      [UNSTABLE] [abi3-ok               ] _moe_C.abi3.so  (stable_shim=0, unstable=82)
      [NO-TORCH] [abi3-ok               ] cumem_allocator.abi3.so
      [NO-TORCH] [abi3-ok               ] spinloop.abi3.so
      [UNSTABLE] [uses-private-api      ] third_party/deep_gemm/_C.cpython-312-x86_64-linux-gnu.so  (stable_shim=0, unstable=59)
      [UNSTABLE] [abi3-ok               ] vllm_flash_attn/_vllm_fa2_C.abi3.so  (stable_shim=0, unstable=85)
      [UNSTABLE] [abi3-ok               ] vllm_flash_attn/_vllm_fa3_C.abi3.so  (stable_shim=0, unstable=81)
```
      
 So 2 unstable extension were removed from _C.abi.so.

### CODE REVIEW

File: csrc/torch_bindings.cpp [hunk 1]
```diff
@@ -152,48 +130,6 @@ TORCH_LIBRARY_EXPAND(TORCH_EXTENSION_NAME, ops) {
...
-      "bool is_scale_transposed) -> ()");
-  ops.impl("rms_norm_per_block_quant", torch::kCUDA, &rms_norm_per_block_quant);
-
-  // Rotary embedding
-  // Apply GPT-NeoX or GPT-J style rotary embedding to query and key.
-  ops.def(
-      "rotary_embedding(Tensor positions, Tensor! query,"
-      "                 Tensor!? key, int head_size,"
-      "                 Tensor cos_sin_cache, bool is_neox, int "
-      "rope_dim_offset=0, bool inverse=False) -> ()");
```
Comment (cleonard530) [inline csrc/torch_bindings.cpp:194]: Merge conflicts location: addition of the "rope_dim_offset=0"

File: csrc/libtorch_stable/pos_encoding_kernels.cu [hunk 2]
```diff
@@ -103,35 +101,37 @@ __global__ void rotary_embedding_kernel(
...
+    std::optional<torch::stable::Tensor> key,
     // null or
     // [batch_size, seq_len, num_kv_heads * head_size] or
     // [num_tokens, num_kv_heads * head_size] or
     // [batch_size, seq_len, num_heads, head_size] or
     // [num_tokens, num_heads, head_size]
     int64_t head_size,
-    torch::Tensor& cos_sin_cache,  // [max_position, rot_dim]
+    torch::stable::Tensor& cos_sin_cache,  // [max_position, rot_dim]
     bool is_neox, int64_t rope_dim_offset, bool inverse) {
```
Comment (cleonard530) [inline csrc/libtorch_stable/pos_encoding_kernels.cu:118]: Merge conflicts location: addition of the "rope_dim_offset"

File: csrc/libtorch_stable/pos_encoding_kernels.cu [hunk 3]
```diff
@@ -166,30 +166,36 @@ void rotary_embedding(
...
-                <<<grid, block, 0, stream>>>(
-                    positions.data_ptr<int64_t>(), query.data_ptr<query_t>(),
-                    key.has_value() ? key->data_ptr<query_t>() : nullptr,
-                    cos_sin_cache.data_ptr<cache_t>(), rot_dim, query_stride,
-                    key_stride, head_stride, num_heads, num_kv_heads, head_size,
-                    rope_dim_offset, inverse);
-          }
-        });
-  });
+  const torch::stable::accelerator::DeviceGuard device_guard(
```
Comment (cleonard530) [inline csrc/libtorch_stable/pos_encoding_kernels.cu:169]: Merge conflicts location: addition of the "rope_dim_offset"

File: csrc/ops.h [hunk 4]
```diff
@@ -67,13 +67,6 @@ void rms_norm(torch::Tensor& out, torch::Tensor& input, torch::Tensor& weight,
...
                         torch::Tensor& weight, double epsilon);
 
-void fused_qk_norm_rope(torch::Tensor& qkv, int64_t num_heads_q,
-                        int64_t num_heads_k, int64_t num_heads_v,
-                        int64_t head_dim, double eps, torch::Tensor& q_weight,
-                        torch::Tensor& k_weight, torch::Tensor& cos_sin_cache,
-                        bool is_neox, torch::Tensor& position_ids,
-                        int64_t forced_token_heads_per_warp);
-
 void fused_deepseek_v4_qnorm_rope_kv_rope_quant_insert(
```
Comment (cleonard530) [inline csrc/ops.h:73]: Merge conflicts location: addition of the "fused_deepseek_v4_qnorm_rope_kv_rope_quant_insert"

File: csrc/torch_bindings.cpp [hunk 5]
```diff
@@ -94,28 +94,6 @@ TORCH_LIBRARY_EXPAND(TORCH_EXTENSION_NAME, ops) {
...
-      "float epsilon) -> ()");
-  ops.impl("fused_add_rms_norm", torch::kCUDA, &fused_add_rms_norm);
-
-  // Function for fused QK Norm and RoPE
-  ops.def(
-      "fused_qk_norm_rope(Tensor! qkv, int num_heads_q, "
-      "int num_heads_k, int num_heads_v, int head_dim, float eps, "
-      "Tensor q_weight, Tensor k_weight, Tensor cos_sin_cache, "
-      "bool is_neox, Tensor position_ids, "
-      "int forced_token_heads_per_warp=-1) -> ()");
```
Comment (cleonard530) [inline csrc/torch_bindings.cpp:116]: Merge conflicts location: addition of the "forced_token_heads_per_warp=-1"

File: csrc/libtorch_stable/fused_qknorm_rope_kernel.cu [hunk 6]
```diff
@@ -18,21 +18,20 @@
 #include <cuda_runtime.h>
 #include <type_traits>
 
-#include <torch/cuda.h>
-#include <ATen/cuda/CUDAContext.h>
-#include <c10/cuda/CUDAGuard.h>
+#include "torch_utils.h"
```
Comment (cleonard530) [inline csrc/libtorch_stable/fused_qknorm_rope_kernel.cu:22]: Merge conflicts location: deletion of `#include <ATen/cuda/CUDAContext.h>`, and 
`#include "async_util.cuh",` two headers that were not there before

File: csrc/libtorch_stable/fused_qknorm_rope_kernel.cu [hunk 7]
```diff
@@ -708,60 +707,64 @@ void launchFusedQKNormRopeNTokenHeads(
...
-    int64_t num_heads_k,      // Number of key heads
-    int64_t num_heads_v,      // Number of value heads
-    int64_t head_dim,         // Dimension per head
-    double eps,               // Epsilon for RMS normalization
-    torch::Tensor& q_weight,  // RMSNorm weights for query [head_dim]
-    torch::Tensor& k_weight,  // RMSNorm weights for key [head_dim]
-    torch::Tensor& cos_sin_cache,  // Cos/sin cache [max_position, head_dim]
-    bool is_neox,                  // Whether RoPE is applied in Neox style
-    torch::Tensor& position_ids,   // Position IDs for RoPE [num_tokens]
-    int64_t forced_token_heads_per_warp  // -1 = auto-select, >0 = forced value
```
Comment (cleonard530) [inline csrc/libtorch_stable/fused_qknorm_rope_kernel.cu:723]: Merge conflicts location: addition of the "forced_token_heads_per_warp"

**Comment by Harry-Chen** [comment]
> The CPU build also uses these declarations

Is it possible that you also migrate these CPU kernels to libtorch stable ABI?

File: csrc/libtorch_stable/fused_qknorm_rope_kernel.cu [hunk 8]
```diff
@@ -708,60 +707,64 @@ void launchFusedQKNormRopeNTokenHeads(
...
-  TORCH_CHECK(
+  STD_TORCH_CHECK(
       qkv.size(1) == total_heads * head_dim,
       "QKV tensor size must match total number of heads and head dimension");
 
-  auto device_id = qkv.get_device();
-  auto stream = at::cuda::getCurrentCUDAStream(device_id);
+  const torch::stable::accelerator::DeviceGuard device_guard(
+      qkv.get_device_index());
+  auto stream = get_current_cuda_stream(qkv.get_device_index());
```
Comment (cleonard530) [inline csrc/libtorch_stable/fused_qknorm_rope_kernel.cu:768]: Merge conflicts location: addition of device_id and stream.

File: csrc/ops.h [hunk 9]
```diff
@@ -99,31 +92,6 @@ void persistent_topk(const torch::Tensor& logits, const torch::Tensor& lengths,
...
-                                      std::optional<torch::Tensor> residual);
-
-void rms_norm_per_block_quant(torch::Tensor& out, torch::Tensor const& input,
-                              torch::Tensor const& weight,
-                              torch::Tensor& scales, double const epsilon,
-                              std::optional<torch::Tensor> scale_ub,
-                              std::optional<torch::Tensor> residual,
-                              int64_t group_size, bool is_scale_transposed);
-
 void silu_and_mul_per_block_quant(torch::Tensor& out,
```
Comment (cleonard530) [inline csrc/ops.h:98]: Merge conflicts location: deletion of #ifdef USE_ROCM

File: csrc/libtorch_stable/quantization/fused_kernels/quant_conversions.cuh [hunk 10]
```diff
@@ -6,7 +6,7 @@
 
 #include "libtorch_stable/quantization/vectorization.cuh"
 // TODO(luka/varun):refactor common.cuh to use this file instead
-#include "../w8a8/fp8/common.cuh"
+#include "../../../quantization/w8a8/fp8/common.cuh"
```
Comment (cleonard530) [inline csrc/libtorch_stable/quantization/fused_kernels/quant_conversions.cuh:9]: Merge conflicts location: `#include "quantization/w8a8/fp8/common.cuh"` was updated to `#include "../w8a8/fp8/common.cuh"` (both replaced by `#include "../../../quantization/w8a8/fp8/common.cuh"`)

**Comment by cleonard530** [comment]
> > The CPU build also uses these declarations
> 
> Is it possible that you also migrate these CPU kernels to libtorch stable ABI?

@Harry-Chen, our plan is to eventually move CPU to stable ABI as well. But first, we want to migrate the CUDA build completely onto to the libtorch stable ABI before we begin with the CPU migration.

**Comment by Harry-Chen** [comment]
> > > The CPU build also uses these declarations
> > 
> > 
> > Is it possible that you also migrate these CPU kernels to libtorch stable ABI?
> 
> @Harry-Chen, our plan is to eventually move CPU to stable ABI as well. But first, we want to migrate the CUDA build completely onto to the libtorch stable ABI before we begin with the CPU migration.

Makes sense. Then could you add some comments to the original header file explaining this?

**Comment by cleonard530** [comment]
@Harry-Chen Also, there was one issue during the merge where I missed an update from TORCH_CHECK to STD_TORCH_CHECK. I have it fixed locally but I doing a quick check to make sure there wasn't another miss before I push the update. I should be able to re-push in a few minutes and will let you know when it's ready!

**Comment by cleonard530** [comment]
> > > > The CPU build also uses these declarations
> > > 
> > > 
> > > Is it possible that you also migrate these CPU kernels to libtorch stable ABI?
> > 
> > 
> > @Harry-Chen, our plan is to eventually move CPU to stable ABI as well. But first, we want to migrate the CUDA build completely onto to the libtorch stable ABI before we begin with the CPU migration.
> 
> Makes sense. Then could you add some comments to the original header file explaining this?

Sure, no problem!

File: csrc/libtorch_stable/fused_qknorm_rope_kernel.cu [hunk 11]
```diff
@@ -604,10 +603,10 @@ void launchFusedQKNormRopeNTokenHeads(
     void const* k_weight, void const* cos_sin_cache, bool const interleave,
     int64_t const* position_ids, int const token_heads_per_warp,
     cudaStream_t stream) {
-  TORCH_CHECK(token_heads_per_warp == 1 || token_heads_per_warp == 2 ||
-                  token_heads_per_warp == 4 || token_heads_per_warp == 8,
-              "token_heads_per_warp must be 1, 2, 4, or 8, got ",
-              token_heads_per_warp);
```
Comment (cleonard530) [inline csrc/libtorch_stable/fused_qknorm_rope_kernel.cu:610]: This block wasn't here before so I had to update TORCH_CHECK.

File: csrc/libtorch_stable/fused_qknorm_rope_kernel.cu [hunk 12]
```diff
@@ -691,7 +690,7 @@ void launchFusedQKNormRopeNTokenHeads(
         });                                                                  \
         break;                                                               \
       default:                                                               \
-        TORCH_CHECK(false, "Unsupported head dimension: ", head_dim);        \
```
Comment (cleonard530) [inline csrc/libtorch_stable/fused_qknorm_rope_kernel.cu:694]: This block wasn't here before so I had to update TORCH_CHECK.

**Comment by cleonard530** [comment]
@Harry-Chen, I added the comments you mentioned. If they look good to you, can you add the 'ready' label?

File: csrc/libtorch_stable/dispatch_utils.h [hunk 13]
```diff
@@ -58,6 +58,35 @@
...
+                      __VA_ARGS__)                                  \
+    THO_DISPATCH_CASE(torch::headeronly::ScalarType::Char, __VA_ARGS__)
+#endif
+
+#define VLLM_STABLE_DISPATCH_QUANT_TYPES(TYPE, NAME, ...) \
+  THO_DISPATCH_SWITCH(TYPE, NAME,                         \
+                      VLLM_STABLE_DISPATCH_CASE_QUANT_TYPES(__VA_ARGS__))
+
+// Group size dispatch (pure C++ if/else, no ATen dependency)
+#define VLLM_STABLE_DISPATCH_GROUP_SIZE(group_size, const_group_size, ...) \
```
Comment (janeyx99) [inline csrc/libtorch_stable/dispatch_utils.h:81]: Outside scope of this PR, but technically this is exactly the same as #define VLLM_DISPATCH_GROUP_SIZE so we could use the stable version everywhere now

**Review by janeyx99 — APPROVED** [review]
Confirmed that this is the same as the PR before, with proper fixes to merge conflicts. LGTM

**Comment by Harry-Chen** [comment]
TIL: [torch-abi-audit](https://pypi.org/project/torch-abi-audit/). Very nice tool!

**Comment by cleonard530** [comment]
> TIL: [torch-abi-audit](https://pypi.org/project/torch-abi-audit/). Very nice tool!

Yes it is! We have @rgommers to thank for that!

**Comment by cleonard530** [comment]
@Harry-Chen I don't think the current errors are due to this PR. Should I try to rebase onto main and try again?

**Comment by cleonard530** [comment]
Also, @Harry-Chen, are you able to give me buildkite access for vllm? This would be helpful in being able to do more with the CI myself, like rerunning failed jobs.

**Comment by rgommers** [comment]
Thanks for the ping, glad you're finding `torch-abi-audit` useful. It's quite new, so if you find gaps or see a need for new features (e.g., there's lots of ways one could do reporting at different levels of detail), please don't hesitate to open issues for them.

**Comment by Harry-Chen** [comment]
> @Harry-Chen I don't think the current errors are due to this PR. Should I try to rebase onto main and try again?

Sure. And if we are stuck on flaky CI tests, we can also ask for a force merge.

> Also, @Harry-Chen, are you able to give me buildkite access for vllm? This would be helpful in being able to do more with the CI myself, like rerunning failed jobs.

CC @khluu on this

**Review by Harry-Chen — APPROVED** [review]
